tsl/src/compression/compression.c

```diff
@@ -340,12 +514,65 @@ compress_chunk(Oid in_table, Oid out_table, const ColumnCompressionInfo **column
 						out_desc->natts,
 						true /*need_bistate*/);
 
-	row_compressor_append_sorted_rows(&row_compressor, sorted_rel, in_desc);
+	if (matched_index_rel != NULL)
+	{
+#ifdef TS_DEBUG
+		const char *compression_path =
+			GetConfigOption("timescaledb.show_compression_path_info", true, false);
+		if (compression_path != NULL && strcmp(compression_path, "on") == 0)
+			elog(INFO, "compress_chunk_indexscan_start");
+#endif
+		index_scan = index_beginscan(in_rel, matched_index_rel, GetTransactionSnapshot(), 0, 0);
+		slot = table_slot_create(in_rel, NULL);
+		index_rescan(index_scan, NULL, 0, NULL, 0);
+		while (index_getnext_slot(index_scan, indexscan_direction, slot))
+		{
+			slot_getallattrs(slot);
+			old_ctx = MemoryContextSwitchTo(row_compressor.per_row_ctx);
+			/* first time through */
+			if (first_iteration)
+			{
+				row_compressor_update_group(&row_compressor, slot);
+				first_iteration = false;
+			}
+			changed_groups = row_compressor_new_row_is_in_new_group(&row_compressor, slot);
+			compressed_row_is_full =
+				row_compressor.rows_compressed_into_current_value >= MAX_ROWS_PER_COMPRESSION;
+			if (compressed_row_is_full || changed_groups)
+			{
+				if (row_compressor.rows_compressed_into_current_value > 0)
+					row_compressor_flush(&row_compressor, mycid, changed_groups);
+				if (changed_groups)
+					row_compressor_update_group(&row_compressor, slot);
+			}
 
-	row_compressor_finish(&row_compressor);
+			row_compressor_append_row(&row_compressor, slot);
+			MemoryContextSwitchTo(old_ctx);
+			ExecClearTuple(slot);
+		}
 
-	tuplesort_end(sorted_rel);
+		run_analyze_on_chunk(in_rel->rd_id);
```

Comment:
why do we want to run analyze on chunk?

Reply:
We need to run analyze to sync the stats. Couple of stats test cases were failing without analyze call.